mitmproxy/net/http/request.py

```diff
@@ -367,13 +367,64 @@ def _get_urlencoded_form(self):
                 pass
         return ()
 
+    def _sanitized_form_data_replacement(self, form_data):
+        # check if unmodified content changes with URL encoding
+        url_decoded_content = urllib.parse.parse_qsl(self.content, keep_blank_values=True)
+        url_reencoded_content = urllib.parse.urlencode(url_decoded_content, doseq=True).encode()
+        if self.content == url_reencoded_content:
+            # encoding the unmodified content doesn't change anything, so we should be good to go on as usual
+            self.content = mitmproxy.net.http.url.encode(form_data).encode()
+        else:
+            # if we are here, we have to alter encoding behavior to mimic the one used in unmodified POST request
+
+            # check if content isn't encoded at all
+            if self.content == urllib.parse.unquote(url_reencoded_content.decode()).encode():
+                # if we land here, the new content mustn't be URL encoded thus we decode the result before setting new content
+                # this is the case for "https://www.facebook.com/ajax/updatestatus.php" but additionally the "=" sign is
+                # omitted on SOME empty keys
+                self.content = urllib.parse.unquote(mitmproxy.net.http.url.encode(form_data)).encode()
+            # check if content isn't encoded at all and equal signs are omitted
+            elif self.content == urllib.parse.unquote(url_reencoded_content.decode()).replace("=&", "&").encode():
+                # remove equal sign for keys without value
+                self.content = urllib.parse.unquote(mitmproxy.net.http.url.encode(form_data)).replace("=&", "&").encode()
+            # the next check should address the facebook issue, which is an absolute weird corner case:
+            # here's a real, but trimmed down data section of a POST request to "https://www.facebook.com/ajax/updatestatus.php"
+            # with content-type 'application/x-www-form-urlencoded'
+            #
+            # self.content:
+            # b'attachment&backdated_date[year]&composer_source_surface=newsfeed&multilingual_specified_lang=&__pc=PHASED%3ADEFAULT'
+            #    * the first two keys have no value, but no equal sign appended
+            #    * the second key shows that no URL encoding ist used ("[" and "]")
+            #    * the third key shows a key=value pair
+            #    * the next key "multilingual_specified_lang" again has no value, but this time an equal sign is appended
+            #    * the last key has assigned a value, which is URL encoded ("%3A" == ":")
```

Comment:
I think we have to live with the fact that we can't provide exact roundtrips in some cases. That's fine, one can still use the raw edit then. This example should go into a test.